python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,213 @@
+    """
+
+    # Lock the assures only one thread at a time can execute the authentication logic.
+    _renew_session_internal_lock = threading.Lock()
+
+    # List of possible states for session renewal.
+    WAITING, CANCELLED, SUCCESS = range(3)
+
+    # When a thread cancels session renewal, this flag is set so other threads know
+    # to raise an exception as well.
+    _auth_state = WAITING
+
+    # Makes access to the thread count and executing logic based on it thread
+    # safe.
+    _renew_session_lock = threading.Lock()
+    # Number of threads who are trying to renew the session.
+    _renew_session_thread_count = 0
+
+    @staticmethod
+    def _renew_session_internal(user, credentials_handler):
+        """
+        Prompts the user for the password. This method should never be called directly
+        and _renew_session should be called instead.
+        :raises AuthenticationCancelled: Raised if the authentication is cancelled.
+        """
+        logger.debug("About to take the authentication lock.")
+        with SessionRenewal._renew_session_internal_lock:
+
+            logger.debug("Took the authentication lock.")
+
+            # When authentication is cancelled, every thread who enter the authentication
+            # critical section should throw as well.
+            if SessionRenewal._auth_state == SessionRenewal.CANCELLED:
+                raise AuthenticationCancelled()
+            # If authentication was successful, simply return.
+            elif SessionRenewal._auth_state == SessionRenewal.SUCCESS:
+                return
+
+            # We're the first thread, so authenticate.
+            try:
+                logger.debug("Not authenticated, requesting user input.")
+                hostname, login, session_token = credentials_handler.authenticate(
+                    user.get_host(),
+                    user.get_login(),
+                    user.get_http_proxy()
+                )
+                SessionRenewal._auth_state = SessionRenewal.SUCCESS
+                logger.debug("Login successful!")
+                user.set_session_token(session_token)
+            except AuthenticationCancelled:
+                SessionRenewal._auth_state = SessionRenewal.CANCELLED
+                logger.debug("Authentication cancelled")
+                raise
+
+    @staticmethod
+    def renew_session(user, credentials_handler):
+        """
+        Prompts the user for the password. This method is thread-safe, meaning if
+        multiple users call this method at the same time, it will keep track of
+        how many threads are currently running inside it and all threads waiting
+        for the authentication to complete will return with the same result
+        as the thread that actually did the authentication, either returning or
+        raising an exception.
+
+        :param user: SessionUser we are re-authenticating.
+        :param credentials_handler: Method that actually prompts the user for
```

Comment:
This looks out of date. Seems like it is called with `credentials_handler.authenticate(...)` above, hinting that the crendetials handler is an object of some sort rather than a method?